### DISCUSSION TITLE
Лента событий изменения состояния контрагентов организации

### DISCUSSION THREAD

**Discussion by JohnSergeev** [diadoc/diadocsdk-1c-docs#1091] [Ideas]
Добрый день.

Общая информация для упрощения «разбора полётов»:
Организация: «ДНС Ритейл»
ИНН: 2540167061
КПП: 254301001
ID организации: [c4390c58315b4323a9b4b5d87d6e868b@diadoc.ru](mailto:c4390c58315b4323a9b4b5d87d6e868b@diadoc.ru)
ID участника ЭДО: 2BM-2540167061-254001001-201312151140099321090

Мы работаем с Диадок с использованием компонент АПИ для 1С
https://diadocsdk-1c.readthedocs.io/ru/latest/index.html
Текущая используемая версия: COM x64 5.51.3.965.

На текущий момент для обработки получения новых документов и обновления существующих документов мы пользуемся лентой событий документооборота – т.е. данными, получаемыми с помощью **GetDocumentEventList()**.

Это – хорошая возможность, т.к. позволяет нам реагировать на изменения т**олько по тем документам, по которым данные изменения происходят**. И освобождает от необходимости перепроверять «вручную» состояние всех имеющихся у нас документов на предмет изменения их статуса, подразделения и т.п.

Однако, когда мы хотим видеть актуальное состояние наших взаимоотношений с контрагентами, нам приходится постоянно проверять состояние **всех контрагентов**, с которыми у нас есть или были какие-то отношения. Вдруг что-то изменилось? А как узнать? Только методами получения данных контрагентов.

И в принципе никакой проблемы и не было бы, если бы у нас не было слишком большого количества контрагентов. Например, только контрагентов, находящихся с нами в установленных отношениях – более 260.000 (двухсот шестидесяти тысяч). А всего контрагентов, находящихся с нами в отношениях, ожидающих одобрения установки отношений с нами или когда-то разорвавших эти отношения – более 440.000. И регулярное обновление информации по состоянию статусов взаимоотношений с этими контрагентами требует очень больших ресурсов. На сервере ЭДО это постоянно «жрёт» память, процессор… И сеть! Мы же эту информацию от вас постоянно запрашиваем и получаем.

При этом реально с течением времени изменяются состояния очень небольшого числа контрагентов. Т.е. на такой большой «куче» контрагентов за, например, 1 час наверняка будут какие-то изменения состояний. Но эти изменения будут касаться ну, например, 100 контрагентов. А может и вообще – 20. Но мы-то проверяем всех! Чуть ли не полмиллиона контрагентов! Т.к. не знаем по какому контрагенту за этот час были изменения, а по какому нет. И число контрагентов только растет.

Эта же проблема касается и структуры подразделений контрагентов – вдруг у какого-то контрагента что-то поменялось в плане подразделений? А как узнать? Только проверив всех. Хотя состояние структуры подразделений меняется ещё реже состояния взаимоотношений – раз эдак в десять!

Т.е. проблема в том, что мы можем получать данные по контрагенту, его подразделениям, состоянию взаимоотношений с ним – всё что угодно, вопросов нет, АПИ позволяет это делать.

НО!
Очень не хватает понимания – а у каких контрагентов вообще хоть что-то из этого поменялось с момента, когда мы последний раз это проверяли?!

Не хватает штуки, которую можно было бы назвать **«Лентой событий по контрагентам организации»**. 

### Т.е. аналога ленты событий документооборота, в которую попадали бы ID контрагентов, по которым с момента последнего проверенного ID события этой ленты были какие-то изменения в статусах отношений, работы или структуры подразделений контрагента.

Было бы здорово, если бы в АПИ появился метод типа **Box.GetCounteragentEventList([AfterEventId])**, возвращающий подобную информацию.

Наличие подобного функционала позволило бы реагировать на изменения состояния только по тем контрагентам, по которым реально что-то поменялось. Это снизило бы как нагрузку на наши сервера, так и нагрузку на ваши, т.к. пользователям АПИ не пришлось бы регулярно проверять состояние по всем своим контрагентам для получения информации о том, вдруг у кого-то что-то поменялось. Плюс, из-за снижения нагрузки и объема проверяемых данных, скорость поддержания состояния контрагентов в учетных системах пользователей АПИ так же бы значительно повысилась. Также это сделало бы затраты на поддержание актуальности состояния этих контрагентов «условно константными», т.к. при любой синхронизации данных обрабатывались бы только те контрагенты, по которым реально что-то поменялось, а не все сразу.

### Вопрос:
Возможна ли реализация ленты событий изменений контрагентов организации? Очень хотелось бы её увидеть. 
**Пользы от такой штуки было бы очень много!**

Спасибо.

**Comment by GilimkhanovDenis** [comment]
Добрый день, @JohnSergeev 

Да. Такую ленту можно добавить - в Диадок появились соответствующие методы. Но в компоненту пока не протягивали

Задача в бэклоге есть. Пока ничего обещать не буду по поводу реализации. но добавим приоритет этой задаче

**Comment by JohnSergeev** [comment]
Добрый день. 
Мы протестировали работу с лентой событий контрагентов в http-версии и обнаружили такой момент:
В ленте отображаются события по контрагентам, НО! **Только в плане смены их статусов!**

Очень бы хотелось, чтобы в ней отображались бы и события, связанные со следующими изменениями:
1. Изменение любых данных по подразделениям контрагента (добавление, удаление, изменение, смена иерархии и т.п.)
2. Изменение любых данных по контрагенту, если такие изменения возможны (например, правка КПП и т.д.)

Без этой информации в ленте событий сложно поддерживать полную актуальность информации по контрагенту, т.к., например, на смену отношений с ним мы отреагируем и обновим информацию. А на удаление у него какого-то подразделения отреагировать не сможем. И потенциально сможем оказаться, например,  в ситуации отправки документа контрагенту в подразделение, которого у него уже нет, а мы у себя информацию об этом не обновили, т.к. просто не знали о этих изменениях.

**Есть ли шанс, что в ленту событий по контрагентам добавят указанные события?**

Нам даже не важна информация о том, что именно с контрагентом произошло (смена статуса отношений или еще что) - главное, чтобы ID контрагента отразился в ленте событий при любом изменении состояния контрагента.

Спасибо.
